### PR TITLE
Support DELETE path

### PR DESCRIPTION
**Pull request by nafraf**
Hi,
This PR implements https://github.com/RedisGraph/RedisGraph/issues/2251

The nodes and edges to be deleted were detected using a similar algorithm of the method:
```
SIValue SIPath_ToList(SIValue p)
```

### CODE REVIEW

File: tests/flow/test_path.py [hunk 1]
```diff
@@ -230,3 +230,23 @@ def test_path_property_access(self):
...
+        node_count = result.result_set[0][0]
+
+        query = """MATCH p = (a:X)-[r:R]-(b:Y) return count(r)"""
+        result = redis_graph.query(query)
+        edge_count = result.result_set[0][0]
+
+        query = """MATCH p = (a:X)-[r:R]-(b:Y) DELETE p"""
+        result = redis_graph.query(query)
+        self.env.assertEquals(result.nodes_deleted, node_count)
+        self.env.assertEquals(result.relationships_deleted, edge_count)
```
Comment (AviAvni) [inline tests/flow/test_path.py:252]: please add last test for empty path
paths with different length

**Review by AviAvni — CHANGES_REQUESTED** [review]
nice work!, thanks

**Comment by AviAvni** [comment]
also can you go over Delete tck tests and remove '@skip' if for the relevant scenarios

File: src/execution_plan/ops/op_delete.c [hunk 2]
```diff
@@ -177,6 +178,23 @@ static Record DeleteConsume(OpBase *opBase) {
...
+			for(size_t i = 0; i < nodeCount - 1 ; i++) {
+				Node *n = Path_GetNode(p, i);
+				array_append(op->deleted_nodes, *n);
+				Edge *e = Path_GetEdge(p, i);
+				array_append(op->deleted_edges, *e);
+			}
+			if(nodeCount > 0) {
+				Node *n = Path_GetNode(p, nodeCount - 1);
+				array_append(op->deleted_nodes, *n);
+			}
```
[the remark this answers is not in the document]
Comment (nafraf) [reply src/execution_plan/ops/op_delete.c]: Thanks. I rewrote the code in [b02a867](https://github.com/RedisGraph/RedisGraph/pull/2567/commits/b02a8673b2d87598822059512e7be4ac416f0107) based on your suggestion, but I did some changes because your code does not get the edge at position 0 and I noticed that the edgeCount variable was not being used.

File: tests/flow/test_path.py [hunk 3]
```diff
@@ -230,3 +230,23 @@ def test_path_property_access(self):
...
+        query = """MATCH (a:X), (b:Y) create (a)-[r:R]->(b)"""
+        redis_graph.query(query)
+
+        query = """MATCH p = (a:X)-[r:R]-(b:Y) return count(a)+count(b)"""
+        result = redis_graph.query(query)
+        node_count = result.result_set[0][0]
+
+        query = """MATCH p = (a:X)-[r:R]-(b:Y) return count(r)"""
+        result = redis_graph.query(query)
+        edge_count = result.result_set[0][0]
```
[the remark this answers is not in the document]
Comment (nafraf) [reply tests/flow/test_path.py]: You are right, I removed that part.

File: tests/flow/test_path.py (hunk 1 above)
Comment (nafraf) [reply tests/flow/test_path.py]: Ok, More test were added in commit [2c74ea3](https://github.com/RedisGraph/RedisGraph/pull/2567/commits/2c74ea32706924c4d71629e222e2054b6fd00faa). 
Please tell me if you think that any other test is missing.

File: src/execution_plan/ops/op_delete.c [hunk 4]
```diff
@@ -10,6 +10,7 @@
 #include "../../query_ctx.h"
 #include "../../graph/graph_hub.h"
 #include "../../arithmetic/arithmetic_expression.h"
+#include "datatypes/path/sipath.h"
```
[the remark this answers is not in the document]
Comment (nafraf) [reply src/execution_plan/ops/op_delete.c:12]: Ok. Done in [b02a867](https://github.com/RedisGraph/RedisGraph/pull/2567/commits/b02a8673b2d87598822059512e7be4ac416f0107)

**Comment by nafraf** [comment]
> also can you go over Delete tck tests and remove '@Skip' if for the relevant scenarios

@AviAvni. Sorry. I didn't find that. Could you explain me with more details? Thanks.

**Review by LiorKogan — CHANGES_REQUESTED** [review]
Please add tests such as
CREATE (a)-[b:B]->(c) 
MATCH p = (d)-[e]-(f) DELETE d,e,f,p

CREATE (a)-[b:B]->(c) 
MATCH p = (d)-[e]-(f) MATCH q = (g)-[h]-(i) DELETE p,q

File: src/execution_plan/ops/op_delete.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (AviAvni) [reply src/execution_plan/ops/op_delete.c]: sorry and this one?
```
if(nodeCount > 0) {
  	Node *n = Path_GetNode(p, 0);
	array_append(op->deleted_nodes, *n);

	for(size_t i = 1; i < nodeCount; i++) {
		Node *n = Path_GetNode(p, i);
		array_append(op->deleted_nodes, *n);
		Edge *e = Path_GetEdge(p, I - 1);
		array_append(op->deleted_edges, *e);
	}
}
```

**Comment by nafraf** [comment]
Tests added in [13febf4](https://github.com/RedisGraph/RedisGraph/pull/2567/commits/13febf45538a44309103649a8187a7e76952e8b1),  this is the current behavior:
```
127.0.0.1:6379> graph.query g "CREATE (a)-[b:B]->(c)"
1) 1) "Nodes created: 2"
   2) "Relationships created: 1"
   3) "Cached execution: 0"
   4) "Query internal execution time: 2.231624 milliseconds"
127.0.0.1:6379> graph.query g "MATCH p = (d)-[e]-(f) DELETE d,e,f,p"
1) 1) "Nodes deleted: 2"
   2) "Relationships deleted: 1"
   3) "Cached execution: 0"
   4) "Query internal execution time: 76.293526 milliseconds"
```

```
127.0.0.1:6379> graph.query g "CREATE (a)-[b:B]->(c)"
1) 1) "Nodes created: 2"
   2) "Relationships created: 1"
   3) "Cached execution: 1"
   4) "Query internal execution time: 0.302370 milliseconds"
127.0.0.1:6379> graph.query g "MATCH p = (d)-[e]-(f) DELETE d,p"
1) 1) "Nodes deleted: 2"
   2) "Relationships deleted: 1"
   3) "Cached execution: 0"
   4) "Query internal execution time: 0.872925 milliseconds"
```

```
127.0.0.1:6379> graph.query g "CREATE (a)-[b:B]->(c)"
1) 1) "Nodes created: 2"
   2) "Relationships created: 1"
   3) "Cached execution: 1"
   4) "Query internal execution time: 0.346143 milliseconds"
127.0.0.1:6379> graph.query g "MATCH p = (d)-[e]-(f) MATCH q = (g)-[h]-(i) DELETE p,q"
1) 1) "Nodes deleted: 2"
   2) "Relationships deleted: 1"
   3) "Cached execution: 0"
   4) "Query internal execution time: 1.209371 milliseconds"
```

File: src/execution_plan/ops/op_delete.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (nafraf) [reply src/execution_plan/ops/op_delete.c]: Ok. It's clearer and works. Change done in [892bebb](https://github.com/RedisGraph/RedisGraph/pull/2567/commits/892bebba360a502c1dd3dd5ec66e1d1d0fa84f83). Thanks

**Comment by nafraf** [comment]
> also can you go over Delete tck tests and remove '@Skip' if for the relevant scenarios

@AviAvni. I tested locally enabling the [Scenario: [1] Detach deleting paths.](https://github.com/RedisGraph/RedisGraph/blob/0e4dea03b984efb67d4fd7e70eea2275e22c7ffe/tests/tck/features/clauses/delete/Delete3.feature#L34) 
The test deletes nodes and relationships as expected, but there is not report of labels deleted as the scenario expects, and test fails.
But running the same test in neo4j sandbox, there is not report of labels deleted.
`Deleted 4 nodes, deleted 3 relationships, completed after 12 ms.`

This is the result with redis-cli:

```
127.0.0.1:6379> GRAPH.QUERY g "CREATE (x:X), (n1), (n2), (n3) CREATE (x)-[:R]->(n1) CREATE (n1)-[:R]->(n2) CREATE (n2)-[:R]->(n3)"
1) 1) "Labels added: 1"
   2) "Nodes created: 4"
   3) "Relationships created: 3"
   4) "Cached execution: 0"
   5) "Query internal execution time: 1.149213 milliseconds"
127.0.0.1:6379> 
127.0.0.1:6379> GRAPH.QUERY g "MATCH p = (:X)-->()-->()-->() DETACH DELETE p"
1) 1) "Nodes deleted: 4"
   2) "Relationships deleted: 3"
   3) "Cached execution: 0"
   4) "Query internal execution time: 1.322558 milliseconds"
```
Do deleted labels should be reported? or maybe is the scenario outdated?
Thanks.

File: src/execution_plan/ops/op_delete.c [hunk 5]
```diff
@@ -177,6 +178,22 @@ static Record DeleteConsume(OpBase *opBase) {
...
+			if(nodeCount > 0) {
+				Node *n = Path_GetNode(p, 0);
+				array_append(op->deleted_nodes, *n);
+
+				for(size_t i = 1; i < nodeCount; i++) {
+					Node *n = Path_GetNode(p, i);
+					array_append(op->deleted_nodes, *n);
+					Edge *e = Path_GetEdge(p, i - 1);
+					array_append(op->deleted_edges, *e);
+				}
```
Comment (AviAvni) [inline src/execution_plan/ops/op_delete.c:194]: sorry after talking with @swilly22 we think it is better to split this to 2 loop to utilize the cpu cache better

```
for(size_t i = 0; i < nodeCount; i++) {
 	Node *n = Path_GetNode(p, i);
 	array_append(op->deleted_nodes, *n);
}
for(size_t i = 0; i < nodeCount - 1; i++) {
 	Edge *e = Path_GetEdge(p, i);
 	array_append(op->deleted_edges, *e);
}
```

Comment (nafraf) [reply src/execution_plan/ops/op_delete.c]: Ok. No problem. Change done

File: src/execution_plan/ops/op_delete.c [hunk 6]
```diff
@@ -177,6 +178,22 @@ static Record DeleteConsume(OpBase *opBase) {
 			array_append(op->deleted_edges, *e);
 			// If evaluating the expression allocated any memory, free it.
 			SIValue_Free(value);
+		} else if(type & T_PATH) {
+			Path *p = (Path *)value.ptrval;
+			size_t nodeCount = Path_NodeCount(p);
+			size_t edgeCount = Path_Len(p);
```
Comment (swilly22) [inline src/execution_plan/ops/op_delete.c:184]:
```c
size_t edgeCount = Path_EdgeCount(p);

for(size_t i = 0; i < nodeCount; i++) {
    Node *n = Path_GetNode(p, i);
    array_append(op->deleted_nodes, *n);
}
for(size_t i = 0; i < edgeCount; i++) {
    Edge *e = Path_GetEdge(p, i);
    array_append(op->deleted_edges, *e);
}
```

**Review by swilly22 — COMMENTED** [review]
@AviAvni @nafraf 
Please see comment, `size_t edgeCount = Path_Len(p);` is wrong.

File: src/execution_plan/ops/op_delete.c (hunk 6 above)
Comment (nafraf) [reply src/execution_plan/ops/op_delete.c:184]: Thanks. I saw that @AviAvni fixed it in c1ad5f6
